testng-core/src/test/java/test/hook/HookableTest.java

```diff
@@ -1,127 +1,211 @@
+    assertThat(listener.getPassedConfigs()).hasSize(4);
+    assertThat(listener.getPassedConfigNames()).contains("bs", "bt", "bc", "bm");
+    SoftAssertions assertions = new SoftAssertions();
+    listener
+        .getPassedConfigs()
+        .forEach(
+            each -> {
+              assertions.assertThat(each.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull();
+              Object[] parameters = (Object[]) each.getAttribute(HOOK_METHOD_PARAMS_ATTRIBUTE);
+              if (parameters != null && parameters.length != 0) {
+                assertions.assertThat(parameters).hasSize(1);
+                assertions.assertThat(parameters[0]).isInstanceOf(Method.class);
+                String methodName = ((Method) parameters[0]).getName();
+                assertions.assertThat(methodName).isEqualTo("hookWasRun");
+              }
+            });
+    assertions.assertAll();
   }
 
-  @Test
-  public void configurableSuccessWithListener() {
-    addClass(ConfigurableSuccessWithListenerTest.class);
-    run();
-
-    Assert.assertEquals(ConfigurableListener.m_hookCount, 4);
-    Assert.assertTrue(ConfigurableSuccessWithListenerTest.m_bs);
-    Assert.assertTrue(ConfigurableSuccessWithListenerTest.m_bt);
-    Assert.assertTrue(ConfigurableSuccessWithListenerTest.m_bc);
-    Assert.assertTrue(ConfigurableSuccessWithListenerTest.m_bm);
-    Assert.assertEquals(ConfigurableListener.m_methodName, "hookWasRun");
+  @DataProvider(name = "getConfigClasses")
+  public Object[][] getConfigClasses() {
+    return new Object[][] {
+      {ConfigurableSuccessSample.class, "IConfigurable as test class"},
+      {ConfigurableSuccessWithListenerSample.class, "IConfigurable as listener"},
+    };
   }
 
   @Test
   public void configurableFailure() {
-    addClass(ConfigurableFailureTest.class);
-    run();
-
-    Assert.assertEquals(BaseConfigurable.m_hookCount, 4);
-    Assert.assertFalse(BaseConfigurable.m_bc);
-    Assert.assertFalse(BaseConfigurable.m_bm);
-    Assert.assertFalse(BaseConfigurable.m_bs);
-    Assert.assertFalse(BaseConfigurable.m_bt);
+    TestNG tng = create(ConfigurableFailureSample.class);
+    TestResultsCollector listener = new TestResultsCollector();
+    tng.addListener(listener);
+    tng.run();
+    assertThat(listener.getPassedConfigNames()).isEmpty();
+    assertThat(listener.getFailedConfig()).hasSize(1);
+    SoftAssertions assertions = new SoftAssertions();
+    listener
+        .getFailedConfig()
+        .forEach(
+            each -> {
+              assertions.assertThat(each.getAttribute(HOOK_INVOKED_ATTRIBUTE)).isNotNull();
+            });
+    assertions.assertAll();
+    // We will now have 3 Skipped configs because when we failed to call the IConfigurable
+    // for the suite level, we explicitly now throw an exception to let users know that they forgot
+    // to call the call back and this upstream config failure causes downstream config skips
+    assertThat(listener.getSkippedConfig()).hasSize(3);
```

Comment:
I'm not sure this behavior change is expected.

Reply:
If a config is not invoked then we are now failing it because the callback was not invoked ( indeterministic behaviour to an explicit behaviour ) hence downstream configs would skip.

Reply:
In fact, it is a good hook if you want any condition on the run of a test.

Reply:
What if the user want to set the result as failed or passed?

Reply:
@juherr - So what are you suggesting? That we leave this indeterministic behaviour as is and assume that a config that is NOT invoked is equivalent to a config that was invoked and passed ?

Reply:
Right, it can be skipped by default if it is not invoked (or it could be a user issue that fails the run).
The question is on the definition of invoked. In my opinion, invoked means "result has a status" and we don't care if the method was called or not (it's a user's decision).

Reply:
@juherr - I spent some more time looking at the implementation. We will NOT be able to make a decision if the result has a status or not because, MethodInvocationHelper when returning back, the status is still `CREATED` which means that there is no way in which we can figure out if the user caused it to be skipped explicitly or if the callback organically passed.

I think its better that we capture this discrepancy as a separate issue. I am going to alter my commit and exclude the changes and have it tracked as a separate issue (See here https://github.com/cbeust/testng/issues/2704)

Reply:
If skipping the test like you did previously is fixing the issue I agree to merge it with a documentation update.

I was just saying that not calling the hook could be a feature because it could be expected in some cases.